ntoskrnl/config/cmparse.c

```diff
@@ -989,40 +1113,731 @@ CmpHandleExitNode(IN OUT PHHIVE *Hive,
+ * @param[in,out] OuterStackArray
+ * A pointer to an array that lives on the caller's stack.
+ * The expected size of the array is up to 32 elements,
+ * which is the imposed limit by CMP_HASH_STACK_LIMIT.
+ * This limit also corresponds to the maximum depth of
+ * subkey levels.
+ *
+ * @param[out] LockedKcbs
+ * A pointer to an array of locked KCBs, returned by the caller.
+ *
+ * @return
+ * Returns STATUS_SUCCESS if all the operations have succeeded without
+ * problems. STATUS_NAME_TOO_LONG is returned if the key pathname has
+ * too many subkey levels (more than 32 levels deep). A failure NTSTATUS
+ * code is returned otherwise. Refer to CmpLookInCache documentation
+ * for more information about other returned status codes.
+ * STATUS_UNSUCCESSFUL is returned if a KCB is referenced too many times.
+ */
+NTSTATUS
+NTAPI
+CmpBuildHashStackAndLookupCache(
+    _In_ PCM_KEY_BODY ParseObject,
+    _Inout_ PCM_KEY_CONTROL_BLOCK *Kcb,
+    _In_ PUNICODE_STRING Current,
+    _Out_ PHHIVE *Hive,
+    _Out_ HCELL_INDEX *Cell,
+    _Out_ PULONG TotalRemainingSubkeys,
+    _Out_ PULONG MatchRemainSubkeyLevel,
+    _Out_ PULONG TotalSubkeys,
+    _Inout_ PULONG OuterStackArray,
+    _Out_ PULONG *LockedKcbs)
+{
+    NTSTATUS Status;
+    ULONG ConvKey;
+    ULONG SubkeysInTotal, RemainingSubkeysInTotal, MatchRemainingSubkeys;
+    CM_HASH_CACHE_STACK HashCacheStack[CMP_SUBKEY_LEVELS_DEPTH_LIMIT];
+
+    /* Make sure it's not a dead KCB */
+    ASSERT((*Kcb)->RefCount > 0);
 
     /* Lock the registry */
     CmpLockRegistry();
 
+    /* Calculate hash values for every subkey this key path has */
+    ConvKey = (*Kcb)->ConvKey;
+    RemainingSubkeysInTotal = CmpComputeHashValue(Current,
+                                                  ConvKey,
+                                                  HashCacheStack,
+                                                  &SubkeysInTotal);
+
+    /* This key path has too many subkeys */
+    if (SubkeysInTotal > CMP_SUBKEY_LEVELS_DEPTH_LIMIT)
+    {
+        DPRINT1("The key path has too many subkeys - %lu\n", SubkeysInTotal);
+        *LockedKcbs = NULL;
+        return STATUS_NAME_TOO_LONG;
+    }
+
     /* Return hive and cell data */
     *Hive = (*Kcb)->KeyHive;
     *Cell = (*Kcb)->KeyCell;
 
-    /* Make sure it's not a dead KCB */
-    ASSERT((*Kcb)->RefCount > 0);
+    /* Do we have any subkeys? */
+    if (!RemainingSubkeysInTotal && !SubkeysInTotal)
+    {
+        /*
+         * We don't have any subkeys nor remaining levels, the
+         * KCB points to the actual key. Lock it.
+         */
+        if (!CmpReferenceKeyControlBlock(*Kcb))
+        {
+            /* This key is opened too many times, bail out */
+            DPRINT1("Could not reference the KCB, too many references (KCB 0x%p)\n", Kcb);
+            return STATUS_UNSUCCESSFUL;
+        }
 
-    /* Reference it */
-    (VOID)CmpReferenceKeyControlBlock(*Kcb);
+        CmpAcquireKcbLockSharedByIndex(GET_HASH_INDEX((*Kcb)->ConvKey));
```

Comment:
Does the Kcb change within this whole CmpBuildHashStackAndLookupCache ? If not, it seems you could reuse that `ConvKey` variable you introduced above and in which you have stored that `(*Kcb)->ConvKey` value.